Modify rule S6491: LaYC format (#2453)
SonarSource/rspec: rules/S6491/cfamily/rule.adoc

> 

Reviewer: I am not sure about not filling in the "How can I fix it" tab. Don't you think the user would need this information every time (s)he sees the issue message?
I think I would, but that may be because I'm not familiar with `std::format`...

Author: Initially, I thought "since the fix is trivial, we don't need it". But, while the fix is only about changing a character or two, it is not always trivial. So now I agree with you.

I added the "how to fix it" section with some more examples. Let me know what you think of it.

`std::format` and the related formatting functions provide two different options to pad numerical values up to a specific width:

* Custom character padding: it applies to many types and can align the text to the left, the center, or the right. +
  For example, `std::format("{:*>5}", num)` aligns `num` to the right (`>`) by inserting as many `*` characters as necessary to reach a width of 5.

> * Numeric padding with `0`: it applies to most arithmetic types by adding `0` before the width specifier. +
>   For example, `std::format("{:05}", num)` adds enough `0` before `num` to align it to the right to reach a width of 5.

Reviewer: The first bullet point describes the outcome of the custom character option while the second bullet point describes how to enable the numeric padding. I did not find it explicit enough so I have asked Island AI to help with that :)
```suggestion
* Numeric padding with `0`: this option is available for most arithmetic types and is enabled by adding `0` before the width specifier. +
  For example, `std::format("{:05}", num)` adds enough `0` before `num` to align it to the right and reach a width of 5.
```

`0` can also be used as a custom character padding.
Not only is the syntax confusing but the results are also surprising when aligning native values to the right:

[source,cpp]
----

> // Noncompliant: the use of character padding results in "00-10"

Reviewer: To make the two comments more consistent
```suggestion
// Noncompliant: character padding results in "00-10"
```

std::format("{:0>5}", -10) // "0" is a custom character padding here.

> // Compliant: use numeric padding to produce "-0010"

Reviewer: For consistency
```suggestion
// Compliant: numeric padding results in "-0010"
```

std::format("{:05}", -10) // The ">" was removed; "0" now means numeric padding.
----

== How to fix it

This rule raises an issue when a format specification uses `0` as a custom padding character to right-align a value instead of using the numeric padding.

=== Code examples

> Numeric padding also works as expected with other specifiers to change the type representation, set the precision, apply locale-specific formatting, etc.
> 
> The following example illustrates the difference in padding behaviors when a sign specifier (such as `{plus}`) and the alternative form (`#`) are used to format the floating-point variable `num`.

Reviewer: I like the ideas of the sentences but I find them a bit hard to follow within the logic of the rule. WDYT about (with the help of Island AI):
```suggestion
This example shows how custom character padding and numeric padding differ when formatting a floating-point variable `num`. It uses a sign specifier (`{plus}`) and the alternative form (`#`).

As a side note, numeric padding also works with other specifiers to change the type representation, set the precision, apply locale-specific formatting, etc.
```

==== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----
auto stringify(float num) {

>   return std::format("{:0>+#5", num); // Custom character padding with "0"

Reviewer: ```suggestion
  return std::format("{:0>+#5", num); // Noncompliant: custom character padding with "0"
```

}
----

==== Compliant solution

A compliant solution is to remove the alignment specifier (`>`) and move the `0` right before the width specifier (`5`) to change its semantics:

[source,cpp,diff-id=1,diff-type=compliant]
----
auto stringify(float num) {
  return std::format("{:+#05}", num); // Numeric padding.
}
----

> Here are some examples of outputs for these two code snippets:

Reviewer: ```suggestion
Here are examples of outputs for the two code snippets:
```
